src/pages/how-does-react-tell-a-class-from-a-function/index.zh-hant.md

```diff
@@ -0,0 +1,553 @@
+  return <p>Hello</p>;
+}
+
+// React 內
+const result = Greeting(props); // <p>Hello</p>
+```
+
+但如果 `Greeting` 是一個類別，React 需要用 `new` 運算符實現出它的實例，*再* 呼叫剛才創建的實例中的 `render` 方法：
+
+```jsx
+// 你的程式碼
+class Greeting extends React.Component {
+  render() {
+    return <p>Hello</p>;
+  }
+}
+
+// React 內
+const instance = new Greeting(props); // Greeting {}
+const result = instance.render(); // <p>Hello</p>
+```
+
+在上述兩種情境中，React 的目的是拿到繪製過的節點（在這例子中，`<p>Hello</p>`），但是確切的步驟取決於 `Greeting` 是如何被定義的。
+
+**所以 React 是如何知道這個東西是類別還是函式？**
+
+正如我 [上一篇文章](/zh-hant/why-do-we-write-super-props/)，**在 React 中，你 *不需要* 知道這個也能具有成效，**我也是多年以來都不知道。請不要把它當成一道面試問題，事實上，這篇文章相較於涉及 React，更多在於涉及 JavaScript。
+
+這個部落格是給想要知道 *為什麽* React 是以這種方式運作，而感到奇心的讀者。你是這個人嗎？那讓我們一起鑽研吧。
+
+**這是一段漫長的旅程，繫好安全帶了。這篇文章並不會涉及太多 React 本身的資訊，相反的，我們會討論 `new`、`this`、`class`、箭頭函式、`prototype`、`__proto__`、`instanceof` 的某些面向，還有這些東西是如何在 JavaScript 中協同運作的。幸運的是，當你 *使用* React 時你不必考慮那麼多，不過如果你正在實做 React 的話…。**
+
+（如果你真的只想知道答案的話，請滾動到最後。）
+
+----
+
+首先，我們需要理解為什麼以不同的方式處理函式跟類別是很重要的。注意我們在呼叫類別時是如何使用 `new` 運算符：
+
+```jsx{5}
+// 如果 Greeting 是個函式
+const result = Greeting(props); // <p>Hello</p>
+
+// 如果 Greeting 是個類別
+const instance = new Greeting(props); // Greeting {}
+const result = instance.render(); // <p>Hello</p>
+```
+
+讓我們粗略地了解 `new` 運算符在 JavaScript 中的作用。
+
+---
+
+早些年 JavaScript 並沒有類別。然而，你可以用函式表達類似類別的模式。具體來說，你可以讓 *任何* 函式扮演類似類別的建構子，藉由在呼叫它前加上 `new`。
+
+```jsx
+// 只是個函式
+function Person(name) {
+  this.name = name;
+}
+
+var fred = new Person('Fred'); // ✅ Person {name: 'Fred'}
+var george = Person('George'); // 🔴 不行
+```
+
+你到今天仍然能寫這樣的程式碼！在 DevTools 中試試看。
+
+如果你呼叫 `Person('Fred')` 時**缺少了** `new`，在這之間的 `this` 會指向某個全球而無用的東西（例如，`window` 或 `undefined`）。所以我們的程式碼將會崩潰，或是做一些像設置 `window.name` 的傻事。
+
+藉由在呼叫前增加 `new`，我們說：「嘿 JavaScript，我知道 `Person` 只是一個函式，但讓我們假裝它就像一個類別的建構子，**創建一個 `{}` 物件並且將 `Person` 函式內部的 `this` 指向這個物件，這樣我就能設置 `this.name` 之類的東西了。然後把這個物件回給我。**」
+
+這就是 `new` 運算符做的事。
+
+```jsx
+var fred = new Person('Fred'); // 與 `Person` 內的 `this` 相同的物件
+```
+
+`new` 運算符也可以創建出我們放在 `Person.prototype` 上的任何東西到 `fred` 物件上：
+
+```jsx{4-6,9}
+function Person(name) {
+  this.name = name;
+}
+Person.prototype.sayHi = function() {
+  alert('Hi, I am ' + this.name);
+}
+
+var fred = new Person('Fred');
+fred.sayHi();
+```
+
+這就是大家在 JavaScript 直接增加類別之前模擬它的方式。
+
+---
+
+所以 `new` 已經在 JavaScript 中存在了一段時間。然而，類別是最近才有的，它讓我們能更貼近我們意圖地重寫上述的程式碼：
+
+```jsx
+class Person {
+  constructor(name) {
+    this.name = name;
+  }
+  sayHi() {
+    alert('Hi, I am ' + this.name);
+  }
+}
+
+let fred = new Person('Fred');
+fred.sayHi();
+```
+
+在程式語言和 API 設計中，*抓住開發者的意圖* 是重要的。
+
+如果你寫一個函式，JavaScript 無法猜測它是否表示該像 `alert()` 呼叫它或是像 `new Person()` 被當成一個建構子對待。忘記對一個函式，如 `Person`， 指定 `new` 會導致令人困惑的行為。
+
+**類別語法讓我們能表示：「這不是一個函式 — 他是一個類別而且有建構子。」**如果你在呼叫它時忘記 `new`，JavaScript 將會舉出一個錯誤：
+
+```jsx
+let fred = new Person('Fred');
+// ✅  如果 Person 是一個函式： 沒問題
+// ✅  如果 Person 是一個類別： 也沒問題
+
+let george = Person('George'); // 我們忘記 `new` 了
+// 😳 如果 Person 是一個像建構子的函式：令人困惑的行為
```

Comment:
I think confusing behavior here means **JavaScript** is confused, not us. So we need to find a consistent way to translate this here on line 162 and line 152.

Reply:
@chloewlin

Reply:
Let's keep it as it is now! I can't think of anything better either.